### PR TITLE
Add m3coordinator capability to downsample using remote m3aggregator

### PR DESCRIPTION
**Pull request by robskillington**
To do:
- [x] Add docker integration test

Followups tracked in "Remote metrics aggregation integration followups" issue https://github.com/m3db/m3/issues/1507.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -208,6 +208,7 @@ docker-integration-test:
 	@./scripts/docker-integration-tests/simple/test.sh
 	@./scripts/docker-integration-tests/prometheus/test.sh
 	@./scripts/docker-integration-tests/carbon/test.sh
+	@./scripts/docker-integration-tests/aggregator/test.sh
```
Comment (prateek) [inline Makefile]: booyah!

File: scripts/docker-integration-tests/aggregator/m3aggregator.yml [hunk 2]
```diff
@@ -0,0 +1,277 @@
+logging:
```
Comment (richardartoul) [inline scripts/docker-integration-tests/aggregator/m3aggregator.yml]: :(

File: scripts/docker-integration-tests/aggregator/test.sh [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+echo "Sending unaggregated carbon metrics to m3coordinator"
+bash -c 'for i in $(seq 1 100); do echo "foo.bar.baz 42 $(date +%s)" | nc 0.0.0.0 7204; sleep 1; done' &
+
+# Track PID to kill on exit
+METRIC_EMIT_PID="$!"
+
+echo "Read back aggregated metric"
+ATTEMPTS=10 TIMEOUT=1 retry_with_backoff read_carbon foo.bar.* 42
```
Comment (richardartoul) [inline scripts/docker-integration-tests/aggregator/test.sh]: Why does this work? is the default aggregation policy "last" or something? Can you add a comment

File: src/cmd/services/m3coordinator/downsample/options.go [hunk 4]
```diff
@@ -175,6 +185,10 @@ type agg struct {
 
 // Configuration configurates a downsampler.
 type Configuration struct {
+	// RemoteAggregator specifies that downsampling should be not
```
Comment (richardartoul) [inline src/cmd/services/m3coordinator/downsample/options.go]: "should be done remotely, not locally."

File: src/cmd/services/m3coordinator/downsample/samples_appender.go [hunk 5]
```diff
@@ -24,20 +24,33 @@ import (
...
+	agg          aggregator.Aggregator
+	clientRemote client.Client
+
 	unownedID       []byte
 	stagedMetadatas metadata.StagedMetadatas
 }
 
 func (a samplesAppender) AppendCounterSample(value int64) error {
+	if a.clientRemote != nil {
+		// Remote client write instead
```
Comment (richardartoul) [inline src/cmd/services/m3coordinator/downsample/samples_appender.go]: . ( I know we need a linter)

File: src/cmd/services/m3coordinator/downsample/samples_appender.go [hunk 6]
```diff
@@ -47,6 +60,15 @@ func (a samplesAppender) AppendCounterSample(value int64) error {
 }
 
 func (a samplesAppender) AppendGaugeSample(value float64) error {
+	if a.clientRemote != nil {
+		// Remote client write instead
```
Comment (richardartoul) [inline src/cmd/services/m3coordinator/downsample/samples_appender.go]: .

File: src/cmd/services/m3coordinator/downsample/samples_appender.go [hunk 7]
```diff
@@ -82,6 +104,14 @@ func (a *samplesAppender) appendTimedSample(sample aggregated.Metric) error {
 					AggregationID: pipeline.AggregationID,
 					StoragePolicy: policy,
 				}
+
+				if a.clientRemote != nil {
+					// Remote client write instead
```
Comment (richardartoul) [inline src/cmd/services/m3coordinator/downsample/samples_appender.go]: .

File: src/cmd/services/m3coordinator/downsample/samples_appender.go [hunk 8]
```diff
@@ -82,6 +104,14 @@ func (a *samplesAppender) appendTimedSample(sample aggregated.Metric) error {
...
+					multiErr = multiErr.Add(a.clientRemote.WriteTimed(sample, metadata))
+					continue
+				}
+
+				// Add timed to local aggregator
```
Comment (richardartoul) [inline src/cmd/services/m3coordinator/downsample/samples_appender.go]: .

File: src/query/config/m3coordinator-local-etcd.yml [hunk 9]
```diff
@@ -28,10 +28,6 @@ clusters:
             - zone: embedded
               endpoints:
                 - 127.0.0.1:2379
-        seedNodes:
```
Comment (richardartoul) [inline src/query/config/m3coordinator-local-etcd.yml]: Why are you removing this? Dont you need it to setup the embedded etcd?

File: scripts/docker-integration-tests/aggregator/docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+      - "7203"
+      - "7204"
+    ports:
+      - "0.0.0.0:7202:7201"
+      - "0.0.0.0:7204:7204"
+    networks:
+      - backend
+    image: "m3coordinator_integration:${REVISION}"
+    volumes:
+      - "./:/etc/m3coordinator/"
```
Comment (richardartoul) [inline scripts/docker-integration-tests/aggregator/docker-compose.yml]: super nit: I generally prefer to mount individual files so its obvious whats needed

File: scripts/docker-integration-tests/m3aggregator.Dockerfile [hunk 11]
```diff
@@ -0,0 +1,12 @@
+FROM alpine:latest AS builder
```
Comment (richardartoul) [inline scripts/docker-integration-tests/m3aggregator.Dockerfile]: Can we open an issue to track pending tasks and include the documentation, but also adding support for this to m3_stack so we can play around with it locally easily

Comment (richardartoul) [reply scripts/docker-integration-tests/m3aggregator.Dockerfile]: Also in the issue we should include that we will need to create a Grafana dashboard

File: src/query/config/m3coordinator-local-etcd.yml (hunk 9 above)
Comment (robskillington) [reply src/query/config/m3coordinator-local-etcd.yml]: Not for the coordinator (this is a coordinator config)

File: scripts/docker-integration-tests/aggregator/m3aggregator.yml (hunk 2 above)
Comment (robskillington) [reply scripts/docker-integration-tests/aggregator/m3aggregator.yml]: Yeah this config is insane.  We’ll have to improve over time.

File: scripts/docker-integration-tests/aggregator/test.sh (hunk 3 above)
Comment (robskillington) [reply scripts/docker-integration-tests/aggregator/test.sh]: Sure can add a comment, the graphite default aggregation policy is average I believe

File: scripts/docker-integration-tests/m3aggregator.Dockerfile (hunk 11 above)
Comment (robskillington) [reply scripts/docker-integration-tests/m3aggregator.Dockerfile]: We already have it in M3 stack actually, can add a dashboard request to an issue for follow ups, let’s do that.

Comment (robskillington) [reply scripts/docker-integration-tests/m3aggregator.Dockerfile]: Opened issue: https://github.com/m3db/m3/issues/1507

**Pull request merged**